### PR TITLE
Add FLAC audio support

### LINKED ISSUE #1190: Add support for FLAC decoding
**Describe the project you are working on:**
Rhythm Game for Desktop

**Describe the problem or limitation you are having in your project:**
The game uses sound tracks to store the audio, and slices everything into keysounds (Samples that are played only when the buttons are pressed), even sounds that won't be displayed as notes (BGM Notes, which are longer), as needed when loading the song.
This is important because the content creators can produce content in a way, that they can intentionally choose to merge sounds that'll play at the same time in one track or not.
For example: Layered Synths, Background Sounds. If they're not merged, it'll be possible for anyone to create harder difficulties in the future. This is to encourage/discourage the creation of new difficulties, remixes and the ability for other people to reuse those sounds for their own creations.

Before this new format it was just a TON of small samples and a BGM file, the flexibility of this new format to organize keysounds in a longer format and the removal of the limit imposed by the nature of the previous format allows for the creators to decide if they want to work the same way as before, export the tracks as is sacrificing flexibility and storage for a faster workflow or to be smart and allow for easier creation of remixes and new content down the line.
Collections can reach hundreds of GBs, this is because of the huge amount of lossless content that is created by the communities through time.
So, with this added flexibility the amount of storage space used per song can be a lot higher than before.
Because of this I think it is really important to use FLAC instead of WAV files to optimize the storage usage from the start.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Decoding of FLAC files through the Engine would fix this issue for this and other games in the genre, allowing to optimize their storage use without the tradeoff of using lossy formats.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
A function that accepts a PoolByteArray as a parameter and returns a decoded PoolByteArray or an instance of AudioStreamSample.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I don't know, it depends on the developer, I'd use it on every project.
And yes, it could be solved with GDNative or Modules but I think there are merits to it being in core.

**Is there a reason why this should be core and not an add-on in the asset library?:**
I honestly think this should be in core because it could benefit not only projects similar to this, but other projects with a sizable Soundtrack to optimize their storage use without sacrificing quality.
It's certainly a handy tool to add to core, and a nod to audiophiles, too.

### PR DESCRIPTION
**Pull request by fire**
Fixes: https://github.com/godotengine/godot-proposals/issues/1190

1. There is a working GDExtension version by DeleteSystem32 [here](https://github.com/DeleteSystem32/godot-flac-extension).
1. dr_flac is a manageable code length.
1. Positive support from Godot Engine users in the godot-proposal for adding to Godot Engine.

---

*EDIT: Added reference to the working GDExtension*

### CODE REVIEW

**Comment by Faless** [comment]
> DeleteSystem32 was unable to create a GDExtension for audio streams due to strict audio code requirements.

Can you expand on this? We definitely should allow custom audio formats to be implemented via GDExtension (even if we decide to include flac support in core).

**Comment by Faless** [comment]
> That was for GDNative, but I've tried reimplementing it as a GDExtension a few days ago and it works fine, [here](https://github.com/DeleteSystem32/godot-flac-extension).

Thanks, I've edited the PR description to link to it.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 237bd0a615df8a0e57bc3d299894abece7b43a0c), it works as expected.

The import options provided are consistent with Ogg Vorbis and MP3, which makes sense as these are designed to be used for music and ambient sounds.

**Testing project:** [test_pr_80160.zip](https://github.com/godotengine/godot/files/12245449/test_pr_80160.zip)[^1]

[^1]: Music in the testing project by [Marc A. Pullen](https://www.youtube.com/@TheMarcAPullen).

FLAC | Ogg Vorbis 412 Kb/s[^2] | MP3 320 Kb/s
-|-|-
Screenshot_20230803_012138 | Screenshot_20230803_012153 | Screenshot_20230803_012147 |

The impact of this PR on binary size is 103 KB (70,337,792 -> 70,440,320 bytes) on a stripped Linux x86_64 release export template (with LTO). This is fairly significant for mobile/web platforms – I wonder if this PR's impact could be reduced somehow.

[^2]: I didn't know Ogg Vorbis could go past 320 Kb/s, but it can when exported from Audacity :slightly_smiling_face:

**Comment by Calinou** [comment]
Note that [dr_flac supports metadata access](https://github.com/mackron/dr_libs/blob/e07e2b8264da5fa1331a0ca3d30a3606084c311f/dr_flac.h#L176-L178), in case we want to implement https://github.com/godotengine/godot-proposals/issues/3307.

File: modules/flac/doc_classes/AudioStreamFLAC.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			            [csharp]
+			            public AudioStreamFLAC LoadFLAC(string path)
+			            {
+			                using var file = FileAccess.Open(path, FileAccess.ModeFlags.Read);
+			                var sound = new AudioStreamFLAC();
+			                sound.Data = file.GetBuffer(file.GetLength());
+			                return sound;
+			            }
+			            [/csharp]
+			            [/codeblocks]
```
Comment (Calinou) [inline modules/flac/doc_classes/AudioStreamFLAC.xml:41]: Remove extraneous space indentation, with the exception of indentation within codeblocks (use tab indentation until the codeblock's level, even within the codeblock).

File: modules/flac/audio_stream_flac.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	drflac *pFlac = nullptr;
+	uint32_t frames_mixed = 0;
+	bool active = false;
+	int loops = 0;
+
+	friend class AudioStreamFLAC;
+
+	Ref<AudioStreamFLAC> flac_stream;
+
+	//void populate_first_frame(int, mp3dec_frame_info_t *);
```
Comment (Calinou) [inline modules/flac/audio_stream_flac.h:60]: Remove this commented out code, as it refers to MP3.

File: modules/flac/doc_classes/AudioStreamFLAC.xml (hunk 1 above)
Comment (fire) [reply modules/flac/doc_classes/AudioStreamFLAC.xml]: May need a check on this.

Comment (raulsntos) [reply modules/flac/doc_classes/AudioStreamFLAC.xml]: It looks good to me now, it has the same indentation as other examples.

File: thirdparty/dr_flac/dr_flac.h [hunk 3]
```diff
@@ -0,0 +1,12532 @@
...
+Build Options
+=============
+#define these options before including this file.
+
+#define DR_FLAC_NO_STDIO
+  Disable `drflac_open_file()` and family.
+
+#define DR_FLAC_NO_OGG
+  Disables support for Ogg/FLAC streams.
+
```
Comment (Calinou) [inline thirdparty/dr_flac/dr_flac.h:200]: Do we support `.ogg` files containing FLAC? If not, we can set this define to reduce binary size a little.

File: thirdparty/dr_flac/dr_flac.h [hunk 4]
```diff
@@ -0,0 +1,12532 @@
...
+APIs can take a long time to initialize because they may need to spend a lot of time finding the first frame.
+
+
+
+Build Options
+=============
+#define these options before including this file.
+
+#define DR_FLAC_NO_STDIO
+  Disable `drflac_open_file()` and family.
```
Comment (Calinou) [inline thirdparty/dr_flac/dr_flac.h:196]: It looks like we don't use `open_file()`, so this define is worth setting to reduce binary size as well.

Comment (fire) [reply thirdparty/dr_flac/dr_flac.h:196]: Both are set here. https://github.com/godotengine/godot/pull/80160/files#diff-b20bb636c4e6135029ad26cd6654f899c6fc96d100b63ea8d222d6e47fc8d34bR34-R36

File: thirdparty/dr_flac/dr_flac.h (hunk 3 above)
Comment (fire) [reply thirdparty/dr_flac/dr_flac.h:200]: Both are set here. https://github.com/godotengine/godot/pull/80160/files#diff-b20bb636c4e6135029ad26cd6654f899c6fc96d100b63ea8d222d6e47fc8d34bR34-R36

**Comment by ellenhp** [comment]
I don't think I support this. I'd guess maybe 0.1% of games will use this, and if it adds 100k that's an enormous dependency to bring on for such a small group of users. There's a working GDExtension for it right? If not, we should figure out how to support it in GDExtension. I think the necessary methods are exposed.

**Comment by DeeJayLSP** [comment]
I'm kinda curious about the decoding efficiency of this compared to Vorbis.

Compared to WAV, FLAC tends to reduce audio file sizes by 34-50%.

In my opinion, one of the few reasons to use FLAC would be to compress sound effects (which are usually in WAV)

If a game has dozens of sound effects, and the performance impact of playing a lot of FLAC sound effects at the same time isn't that big (needs to be tested), then 103KB shouldn't be something to worry about.

**Comment by ellenhp** [comment]
If a game needs this they can use the GDExtension.

I know firsthand (trust me :laughing:) that it's very frustrating when a feature you need isn't getting merged, but keeping the engine small is worth the headache. There's some background here: https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell/

A hundred kilobytes is a lot for a feature like this, and if there's a working GDExtension I don't see any reason to merge it. I think GDExtension's whole reason for existing is to support niche use cases like this.

**Comment by ParadoxV5** [comment]
Are we certain that FLAC is niche? It’s both popular enough and underütilized to me.

**Comment by Calinou** [comment]
> Are we certain that FLAC is niche? It’s both popular enough and underütilized to me.

In the context of game development, it is considered a niche format as you don't benefit from lossless audio that much:

- In a game, audio often plays with other samples at the same time, so it's usually indistinguishable from high-quality lossy audio. It's hard to appreciate music quality at its fullest when there are sound effects, voices and ambient sounds playing at the same time.
- High-quality lossy audio (such as Ogg Vorbis at 320 Kb/s) is already not distinguishable for most users, while being significantly smaller in size.
- It's even less noticeable for users using wireless audio solutions, especially Bluetooth with its audio stack being subpar on desktop platforms. High-quality Bluetooth audio is only available on mobile platforms, and not without some fiddling.

**Comment by DeeJayLSP** [comment]
For comparison, I did a quick release build on DeleteSystem32's implementation.

The stripped lib sizes are the following:

| Platform         | Size       |
|:-----------------|:-----------|
| Windows (MinGW)  | 3034.00 KB |
| Linux            | 4443.84 KB |

**Comment by DeeJayLSP** [comment]
> The impact of this PR on binary size is 103 KB (70,337,792 -> 70,440,320 bytes) on a stripped Linux x86_64 release export template (with LTO). This is fairly significant for mobile/web platforms

Out of curiosity, I decided to do a minimal release production build for the web (some modules ended up being built, but the difference is what matters more here).

The command used was `scons -j$(nproc) platform=web target=template_release production=yes <disable_most_things_according_to_docs_optimizing_for_size>`, with and without flac enabled.

The resulting *.wasm* file got increased by 58.48KB (25,345,358 to 25,403,838 bytes).

The resulting *.zip* file got increased by 19.8KB (6,036,845 to 6,056,647).

One thing to be aware of is that `dr_flac` includes SIMD optimizations for Desktop and Mobile platforms. Therefore, this decrease is likely not valid for mobile.

**Comment by ellenhp** [comment]
Godot is not audiophile software, so I don't personally see the need to take on even 1k of extra binary size for audiophile codecs considering the fact that there is a working GDExtension for games that require it. Is there anyone out there who absolutely needs this to be in core? I've seen a few people say FLAC support would benefit them. Maybe one of those people could explain why the GDExtension doesn't satisfy their needs? This feels hypothetical right now and we don't generally merge PRs that solve hypothetical problems.

I know I posted this earlier but useful context for anyone skimming: https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell/

**Comment by DeeJayLSP** [comment]
It isn't just for audiophile purposes.

In context of sound effects, using formats like MP3 or Vorbis isn't really good for when someone wants to play lots of sound effects at the same time. While FLAC doesn't decrease the file size at the same level, it is capable of decoding about 1.4-1.5x faster (at least that's what libFLAC on ffmpeg shows when decoding to PCM) therefore being less demanding than those.

This is a possible use case for FLAC. A game with lots of sound effects (which certainly isn't just a few cases) could benefit by having those as FLAC instead.

For that context, I don't think a GDExtension helps when it's over 30x larger than a built-in implementation.

**Comment by ellenhp** [comment]
Some disorganized thoughts.

1. I stand by my statement that FLAC is an audiophile format, for reasons to follow:
2. If you're playing lots of sound effects, the mix is going to be very muddy. It doesn't sound good to play lots of sound effects at once. We do want to support this use-case anyway because sometimes it's really funny to play a game that plays a bunch of sfx at once, and we definitely don't want underruns ever, even if the sound ends up being unpleasant to listen to. But this is a thing to consider when we're talking about audiophile codecs where quality is the primary concern (which I will back up with some data at the end).
3. If a game needs lots of overlapping sounds without it being intelligible which sounds are which (humans can't keep track of more than like 10-20 different sounds at once IIRC) I think it's common practice to combine them in a DAW, because then the sound engineer can equalize things, use compressors to increase intelligibility, pitch shift, stretch individual sounds in time, and generally just create a nice chaotic sound. You can't do this at runtime.
4. Using FLAC for short sound effects doesn't generally decrease their size as far as I can tell, and sometimes *increases* their size. I tested several different sound effects distributed as WAVs. I'm not sure of the licensing so I'm not going to post these specific sfx, but I'd love to see testing by others on this too. Results at the end of this comment.
5. The break-even point on compressing a sound effect with FLAC purely in terms of size seems to be about 5-10 seconds long at 11k sample rate. Playing more sounds than can be decoded by vorbis on a given system, each of which is about 5-10 seconds long is going to result in an absolute cacophony. It will sound terrible. If someone doesn't mind things sounding terrible, or if they want that, and they can't overlay the sounds in a DAW, they should use ADPCM wavs. There was some talk about removing this but reduz didn't want to because it's very light on binary size and has zero maintenance burden. In retrospect that makes a lot of sense.
6. I think libFLAC does allocation during decoding based on what I've seen [here](https://gitlab.xiph.org/xiph/flac/-/issues/1), which we have a strict rule against to the point that we even went to the trouble of avoiding the use of libogg on the audio thread :grimacing:  I don't think the rule should be as strict as it is, but that's a discussion for another time.

I'm sorry about the wall of text and I promise I'm not trying to be difficult, I just don't think this is a useful feature for most people who use Godot. The main benefit here seems to be compatibility with user-provided sounds. I know someone is working on a DAW in Godot and this seems like a good feature for them, but I think for that application a GDExtension is probably fine?

```Input #0, wav, from 'car_crash2.wav':
  Duration: 00:00:04.96, bitrate: 88 kb/s
  Stream #0:0: Audio: pcm_u8 ([1][0][0][0] / 0x0001), 11025 Hz, 1 channels, u8, 88 kb/s
-rw-r--r--. 1 ellen ellen 54297 Aug 28 17:56 car_crash2.flac
-rw-r--r--. 1 ellen ellen 23438 Aug 28 17:57 car_crash2.ogg
-rw-r--r--. 1 ellen ellen 54701 Aug 28 17:55 car_crash2.wav
```

```Input #0, wav, from 'aaa.wav':
  Duration: 00:00:02.13, bitrate: 89 kb/s
  Stream #0:0: Audio: pcm_u8 ([1][0][0][0] / 0x0001), 11127 Hz, 1 channels, u8, 89 kb/s
-rw-r--r--. 1 ellen ellen 23846 Aug 28 17:52 aaa.flac
-rw-r--r--. 1 ellen ellen 11492 Aug 28 17:52 aaa.ogg
-rw-r--r--. 1 ellen ellen 23744 Aug 28 17:52 aaa.wav
```

```Input #0, wav, from 'carpentry.wav':
  Duration: 00:00:09.13, bitrate: 88 kb/s
  Stream #0:0: Audio: pcm_u8 ([1][0][0][0] / 0x0001), 11025 Hz, 1 channels, u8, 88 kb/s
-rw-r--r--. 1 ellen ellen  78687 Aug 28 17:55 carpentry.flac
-rw-r--r--. 1 ellen ellen  39876 Aug 28 17:55 carpentry.ogg
-rw-r--r--. 1 ellen ellen 100732 Aug 28 17:55 carpentry.wav
```

**Comment by DeeJayLSP** [comment]
Out of curiosity, I decided to do the same test with a pack of 51 UI sound effects from Kenney, which does include some really short sound effects.

In total, the total WAV size got to 2.3MiB, while the total FLAC size got to 1.1MiB.

I can confirm, however, that some sound effects increased in size upon conversion. Those are *really* short effects (less than 0.1s duration).

**Comment by ellenhp** [comment]
> In total, the total WAV size got to 2.3MiB, while the total FLAC size got to 1.1MiB.

Curious if you could try IMA-ADPCM. In Godot it's an import option and you'd just normal use the WAV. In ffmpeg you can just add `-acodec adpcm_ima_wav`.

I bet it will be substantially smaller than FLAC, so if your primary concern is some combination of speed and size it will outperform FLAC in both, while sounding mediocre. But again, if your mix is muddy that really isn't going to matter IMO.

**Comment by Calinou** [comment]
If you need a better alternative to IMA-ADPCM, [QOA](https://qoaformat.org/) could be worth considering.

**Comment by akien-mga** [comment]
> For that context, I don't think a GDExtension helps when it's over 30x larger than a built-in implementation.

For the record, that's a known godot-cpp bug which will be solved eventually. https://github.com/godotengine/godot-cpp/issues/1160

We shouldn't hyperfocus on size as the sole deciding factor anyway. 1k or 100k or even a megabyte can totally be justified for a feature that's going to be used by most Godot users. For a niche feature, the size matters but what's a lot more important is the maintenance burden, compilation time, etc. "Death by a thousand cuts" impacts both binary size and compilation time, and the more thirdparty libraries we add, the more we need to be aware of security vulnerabilities reported against them, or even just bugfixes we need to bring downstream.

The actual use case for FLAC doesn't seem to have been well identified yet. I see engineers reasoning about whether it makes sense or not from a size or a decompression time point of view, but I don't see yet sound designers telling us that they're using FLAC in their professional pipeline and would need it in Godot.

**Comment by ParadoxV5** [comment]
What roles do audio quality play in Music/Rhythm games? 🤔

**Comment by YYF233333** [comment]
> What roles do audio quality play in Music/Rhythm games? 🤔

Commercial rhythm games typically use ogg just as other games(take a look at arcaea), however community rhythm games definitely need FLAC support as they support import chart pack from outside.

In my opinion audio quality doesn't matter that much, if a beat can only be recognized in lossless format, then it is just too soft to add a key on it. If you just want to listen the music you can always go and download a FLAC version.

**Comment by DeeJayLSP** [comment]
This implementation outputs a glitchy audio if the file has any channels but 2.

**Comment by DeeJayLSP** [comment]
Coming here two months later just to say: forget anything I said about FLAC for sound effects.

Turns out FFmpeg decodes FLAC with multithreading by default (as phoboslab claims). And I ended up testing exactly like that.

QOA is a better format than FLAC except in quality, and better than IMA-ADPCM except in decoding speed, but both are reasonably good enough.

**Comment by YuriSizov** [comment]
@DeeJayLSP Should we close this PR then as you've reached a consensus that FLAC support is not valuable in core?

**Comment by YuriSizov** [comment]
Okay then. Given the concerns from the audio team members as well, I'm going to close this issue. But I hope to see a GDExtension for FLAC for those who need it 🙂

**Comment by ParadoxV5** [comment]
There are two topics in this PR thread.
The more prominent one is rather whether FLAC is worthy of joining Core and if alternative options are more suitable.
We should rather host this discussion in the proposal – godotengine/godot-proposals#1190. I didn’t read it, just thought the consensus is positive because this PR exist*ed*.

That said, for this PR specifically, the implementation isn’t great. Over a hundred KBs, yet it’s still only half-baked.

> This implementation outputs a glitchy audio if the file has only one channel, and won't play at all if it's 3 or above;

Even if we deem FLAC Core-worthy, perhaps another better implementation can make this PR obsolete.

**Comment by fire** [comment]
> Fixed an issue that caused single channel MP3 files to play badly.

@DeeJayLSP What was the fix?

**Comment by DeeJayLSP** [comment]
@fire `drflac_read_pcm_frames_f32` decodes directly to F32 sample data, without caring about the number of channels. And the buffer we're passing to it is always stereo, which means something like this will happen if the audio is mono:

AudioFrame[f1, f2]
AudioFrame[f3, f4]
AudioFrame[f5, f6]
AudioFrame[f7, f8]

When it should be:

AudioFrame[f1, f1]
AudioFrame[f2, f2]
AudioFrame[f3, f3]
AudioFrame[f4, f4]

The fix consists on declaring a fixed-size buffer (let's call it `buf_frame`), reading 1 frame at a time using that buffer, then manually assign the decoded samples on `buf_frame` to the corresponding p_buffer index (always get `buf_frame[0]` for left channel, then perform an operation to find what should be provided to the right channel.

You can essentially do the same as the `dr_mp3` implementation, just swap everything to the corresponding `dr_flac` functions.

**Pull request closed without merging**